### ISSUE TITLE
ValueError: ('Unrecognized keyword arguments:', dict_keys(['ragged']))

### ISSUE THREAD

**Issue by AndreasTraut** [fnandocontreras/mask-detector#1]
When I try to run the mask-detector I get the following error message: 

`ValueError: ('Unrecognized keyword arguments:', dict_keys(['ragged']))`

The complete log is [python app.txt](https://github.com/fnandocontreras/mask-detector/files/5500245/python.app.txt) or further below in this issue message. 

Can you help on this? 

I followed your installation instructions ("conda env create -f environment.yml", "conda activate maskdetector") and am not sure, if something is goring wrong here already. See here [conda activate maskdetector.txt](https://github.com/fnandocontreras/mask-detector/files/5500246/conda.activate.maskdetector.txt)). 

The complete log is as follows: 
(maskdetector) C:\Users\andre\Documents\GitHub\mask-detector>python app.py
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\framework\dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint16 = np.dtype([("quint16", np.uint16, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint32 = np.dtype([("qint32", np.int32, 1)])
C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorboard\compat\tensorflow_stub\dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  np_resource = np.dtype([("resource", np.ubyte, 1)])
Using TensorFlow backend.
Traceback (most recent call last):
  File "app.py", line 22, in <module>
    model = load_model('model.h5')
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\saving\save.py", line 146, in load_model
    return hdf5_format.load_model_from_hdf5(filepath, custom_objects, compile)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\saving\hdf5_format.py", line 212, in load_model_from_hdf5
    custom_objects=custom_objects)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\saving\model_config.py", line 55, in model_from_config
    return deserialize(config, custom_objects=custom_objects)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\layers\serialization.py", line 89, in deserialize
    printable_module_name='layer')
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\utils\generic_utils.py", line 192, in deserialize_keras_object
    list(custom_objects.items())))
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\engine\sequential.py", line 352, in from_config
    custom_objects=custom_objects)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\layers\serialization.py", line 89, in deserialize
    printable_module_name='layer')
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\utils\generic_utils.py", line 192, in deserialize_keras_object
    list(custom_objects.items())))
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\engine\network.py", line 1121, in from_config
    process_layer(layer_data)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\engine\network.py", line 1105, in process_layer
    layer = deserialize_layer(layer_data, custom_objects=custom_objects)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\layers\serialization.py", line 89, in deserialize
    printable_module_name='layer')
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\utils\generic_utils.py", line 194, in deserialize_keras_object
    return cls.from_config(cls_config)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\engine\base_layer.py", line 446, in from_config
    return cls(**config)
  File "C:\Users\andre\Anaconda3\envs\maskdetector\lib\site-packages\tensorflow\python\keras\engine\input_layer.py", line 80, in __init__
    raise ValueError('Unrecognized keyword arguments:', kwargs.keys())
ValueError: ('Unrecognized keyword arguments:', dict_keys(['ragged']))

**Comment by fnandocontreras** [comment]
Hello, sorry for the delay. Taking a look at your logs, this error suggest me this is a tensorflow/keras deserialization issue. It might be caused by a version mismatch. (a different version of tensorflow/keras between the environment where the model was serialized (mine) and the environment where the model is being deserialized (yours)). Could you please let me know what your versions of tensorflow/keras are?

**Comment by fnandocontreras** [comment]
you can check this by typing in your commandline:

`conda list -n maskdetector`
mines are: 
tensorflow            2.1.0
keras                     2.3.1

a workaround if the problem relies on different versions might be to retrain the model, you can use this jupyter notebook to do so:
https://github.com/fnandocontreras/mask-detector/blob/master/training_model.ipynb

**Comment by AndreasTraut** [comment]
Thanks for you reply. I followed your instructions. The following works:    
`git clone https://github.com/fnandocontreras/mask-detector.git`
And also creating an environmoent:  
`conda env create -f environment.yml`
But the next command runs into a lot of errors (see the complete log [here](https://github.com/fnandocontreras/mask-detector/files/5500246/conda.activate.maskdetector.txt)): 
`conda activate maskdetector`
![grafik](https://user-images.githubusercontent.com/55921277/101329819-2c901780-3872-11eb-868d-e4c4952fd96b.png)

I also followed your hint: 
`conda list -n maskdetector`
and I have tensorflow 1.14.0 and keras 2.3.1

**Comment by AndreasTraut** [comment]
I have another environment named "tf2" on my computer, where I have tensorflow 2.1.0 and keras 2.4.3 installed, but your mask-detector doesn't work here either in this environment. I installed mtcnn into my tf2-enviroment (`pip install mtcnn`) but then I get more errors, when I want to run your mask-detector app `python app.py`. I get into troubles with dependencies of already installed packages of my "tf2" enviroment and your requirements. I don't want to destroy my "tf2" enviroment so I will stop adding more packages here.

**Comment by fnandocontreras** [comment]
looking at you error logs when trying to activate the environment i see that your keras package is setting the backend to theano with: set "KERAS_BACKEND=theano", 
Please refer to this issue on Keras github: https://github.com/keras-team/keras/issues/6925
for some reason that i ignore your conda environment is setting the keras backend to theano instead of tensorflow. 
Let me know if the above link to the issue was helpful

**Comment by fnandocontreras** [comment]
Another workaround might be to install the dependencies with pip like:
`pip install numpy tensorflow keras mtcnn pillow opencv`
but you have to be sure to run the program outside of conda environment, if you have python installed in your PC, using the classic CMD (not anacoda prompt) should work

**Comment by AndreasTraut** [comment]
I did the following on your two comments: 
**First:** the theano backend. 
I went to the directory `C:\Users\andre\Anaconda3\envs\maskdetector\etc\conda\activate.d\` and saw the `keras_activate.bat` file there. I found `set "KERAS_BACKEND=theano"` there and changed it to `set "KERAS_BACKEND=tensorflow"`. After having done this I tried activate maskdetector, and got these messages: 
[activate maskdetector.txt](https://github.com/fnandocontreras/mask-detector/files/5677943/activate.maskdetector.txt)
Your app didn't work by typing `python app.py`
Then I deleted the `keras_activate.bat` file and another bat (`vs2017_compiler_vars.bat`) as your github link suggested. 
Same here: your app didn't work. 

**Second:** the pip install
I installed `pip install numpy tensorflow keras mtcnn pillow opencv` and found [here](https://stackoverflow.com/questions/37776228/pycharm-python-opencv-and-cv2-install-error) that it should be `install opencv-python` instead of `install opencv`. 
After having done this I tried starting your app with `python app.py` again but got these messages: 
[python app.py .txt](https://github.com/fnandocontreras/mask-detector/files/5677966/python.app.py.txt)
```
model_config = json.loads(model_config.decode('utf-8'))
AttributeError: 'str' object has no attribute 'decode'

```
**Third:** conda command vs cmd
When I use the cmd instead of the conda command it says, that `cv2 `is not found
[cv2 not found.txt](https://github.com/fnandocontreras/mask-detector/files/5677981/cv2.not.found.txt)
I changed `import cv2` to different other things, but am not able to run the app. 

Sorry, I am not a specilist in these installation routines and environments. It seems to be more difficult than expected and even the Anaconda distribution is not really helpful here.
